pkg/scheduler/allocator.go

```diff
@@ -214,6 +211,56 @@ func (m *Scheduler) tryBatchAllocation(partition string, candidates []*Schedulin
     return allocations, failedAsks
 }
 
+// TODO: convert this as an interface.
+func evaluateForSchedulingPolicy(m *Scheduler, nodes []*SchedulingNode, partition string,
+    candidate *SchedulingAllocationAsk, partitionContext *PartitionSchedulingContext) []*SchedulingNode {
+
+    // If bin-packing is not enabled, simply return
+    if !partitionContext.partition.NeedBinPackingSchedulingPolicy() {
+        // Sort by MAX_AVAILABLE resources.
+        // TODO, this should be configurable.
+        SortNodes(nodes, MaxAvailableResources)
+        return nodes
+    }
+
+    // Do an in-place sorting
+    nodes = m.SortAllNodesWithAscendingResource(partition)
+
+    for i := 0; i < len(nodes); i++ {
+        node := nodes[i]
+        if !node.CheckAllocateConditions(candidate.AskProto.AllocationKey) {
+            // skip the node if conditions can not be satisfied
+            continue
+        }
+
+        // once we have a node which can fit the allocation ask, send back for scheduling
+        if node.CheckAndAllocateResource(candidate.AllocatedResource, false /* preemptionPhase */) {
+            if log.IsDebugEnabled() {
+                log.Logger().Debug("Selected node:",
+                    zap.Any("id", node))
+            }
+            return append(make([]*SchedulingNode, 0, 1), node)
+        }
+    }
```

Comment:
Why handle checkAllocationConditition, and check Allocate resource here? We have logic in the `allocate` already. I think we should not let the "node sorting policy" to handle allocate or not based on a specific resource request. 

The node sort policy should just sort the node, it may filter out some invalid node (such as 0 available capacity).         And I think we should not create a copy of node here. 

Ideally, we should have a state kept in node-sort-policy, when new node updated, new pod allocated, etc. it will be notified instead of sorting every time.

Reply:
done